lib/segment/src/index/field_index/full_text_index/inverted_index/mod.rs

```diff
@@ -86,6 +92,33 @@ impl Document {
     pub fn len(&self) -> usize {
         self.0.len()
     }
+
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    pub fn tokens(&self) -> &[TokenId] {
+        &self.0
+    }
+
+    pub fn to_token_set(&self) -> TokenSet {
+        self.0.iter().copied().collect()
+    }
+
+    /// Checks if the current document contains the given phrase.
+    ///
+    /// Returns false if the phrase is empty
+    pub fn has_phrase(&self, phrase: &Document) -> bool {
+        let doc = self.0.as_slice();
+        let phrase = phrase.0.as_slice();
+
+        if doc.is_empty() {
+            return false;
+        }
+
+        // simple check for tokens in the same order as phrase
+        doc.windows(phrase.len()).any(|window| window == phrase)
```

Comment:
We can potentially benefit from something like [KMP](https://en.wikipedia.org/wiki/Knuth%E2%80%93Morris%E2%80%93Pratt_algorithm) here / search from the back and jump the whole length forward on mismatch. I understand it's tokens here, but the same applies.

We could even implement it in a generic way because I believe we use it in more places.

Just noting it here as potential side quest :nerd_face: 

Edit: as far as I understand this can be quite hot, so it might make sense anyway.

Reply:
It is a potential side-quest, yes 😄 . 

AFAIK KMP works best when the needle and the haystack are long, which may not necessarily be the case when it is already-tokenized. Moreover, it adds overhead that is not justified for small sequences.